once/_iterator_wrappers.py

```diff
@@ -154,8 +158,14 @@ def yield_results(self) -> collections.abc.Generator:
             except StopIteration:
                 with self.lock:
                     self.finished = True
+                    self.generating = False
                     self.generator = None  # Allow this to be GCed.
+            except Exception as e:
+                with self.lock:
+                    self.finished = True
                     self.generating = False
+                    self.exception = e
```

Comment:
```suggestion
                    # We need to keep track of the exception so that we can raise it in the same
                    # position every time the iterator is called.
                    self.exception = e
```

Reply:
done